Vars and namespaces with specified metadata keys can be excluded

lread opened the pull request:
I made an attempt to not affect current performance when
:exclude-meta option is not specified.

To support this change:
- extended our test-resources example
- upgraded to current clj-kondo (for meta analysis data)
- added command line support for repeating options and keyword values
- because command line parsing is more complex, tests now invoke
  api-diff/-main instead of api-diff/api-diff

Also:
- added the most minimal support for failing on bad cmd line options

Closes #4

lread committed:
Vars and namespaces with specified metadata keys can be excluded

I made an attempt to not affect current performance when
:exclude-meta option is not specified.

To support this change:
- extended our test-resources example
- upgraded to current clj-kondo (for meta analysis data)
- added command line support for repeating options and keyword values
- because command line parsing is more complex, tests now invoke
  api-diff/-main instead of api-diff/api-diff

Also:
- added the most minimal support for failing on bad cmd line options

Closes #4

lread commented:
Ha! Result output order is not deterministic! Will fix.

lread committed:
Results are now explicitly sorted by namespace then row

lread commented:
Ha! Windows file separators! Will fix!

lread committed:
Adjust tests to deal with file path separator OS diffs

Includes work-around for https://github.com/clj-kondo/clj-kondo/issues/1438

borkdude commented on src/borkdude/api_diff.clj:
```diff
@@ -57,17 +88,28 @@
-    (api-diff (update-some opts :lib symbol))))
+  (let [{:keys [lib v1 v2 path1 path2] :as opts}
+        (parse-opts args {:exclude-meta {:multi true
+                                         :type :keyword}
```
@lread I think instead of hardcoding keywords -> functions or multi into the parse-fn we can make this more general.

I propose 

- `:collect-fn`: a function which collects arguments into a collection. Defaults to picking the last argument (`identity`). Can be `(fnil conj [])` or `(fnil conj #{})` for example for collecting into vector or set, or assoc-ing into a map, whatever.
- `:parse-fn`: the value is parsed through parse-fn, defaults to `identity`, but can be `keyword`, `symbol`, `#(str/split % ",")` or whatever.

lread replied:
Yes! I think we chatted about such an approach, but I did not deliver that, did I? 🙂 
Happy to give that a shot.

borkdude replied:
@lread While at it, I think the map from cmd line keywords to options might need one more level of nesting to avoid conflicts with other future options, although we don't really need to worry about it now since the function isn't public.

lread replied:
Yeah, since it is private, I think it can evolve as needed?

borkdude replied:
Sure, but let's do `:parse-fn` etc now.

lread replied:
Agreed, will do!

lread committed:
review feedback: rework cmd line option parsing

Now using :parse-fn and :collect-fn instead of an
app-specific-dsl-ish strategy.
